group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java

```diff
@@ -190,6 +216,127 @@ boolean canTransitionFrom(CoordinatorState state) {
+     * {@link TimeoutOperation} operations scheduled by the coordinators.
+     *
+     * It also keeps track of all the scheduled {@link TimerTask}. This allows timeout operations to be
+     * cancelled or rescheduled. When a timer is cancelled or overridden, the previous timer is guaranteed to
+     * not be executed even if it already expired and got pushed to the event processor.
+     *
+     * When a timer fails with an unexpected exception, the timer is rescheduled with a backoff.
+     */
+    class EventBasedCoordinatorTimer implements CoordinatorTimer<U> {
+        /**
+         * The logger.
+         */
+        final Logger log;
+
+        /**
+         * The topic partition.
+         */
+        final TopicPartition tp;
+
+        /**
+         * The scheduled timers keyed by their key.
+         */
+        final Map<String, TimerTask> tasks = new HashMap<>();
+
+        EventBasedCoordinatorTimer(TopicPartition tp, LogContext logContext) {
+            this.tp = tp;
+            this.log = logContext.logger(EventBasedCoordinatorTimer.class);
+        }
+
+        @Override
+        public void schedule(
+            String key,
+            long delay,
+            TimeUnit unit,
+            TimeoutOperation<U> operation
+        ) {
+            // The TimerTask wraps the TimeoutOperation into a CoordinatorWriteEvent. When the TimerTask
+            // expires, the event is push to the queue of the coordinator runtime to be executed. This
+            // ensure that the threading model of the runtime is respected.
+            TimerTask task = new TimerTask(unit.toMillis(delay)) {
+                @Override
+                public void run() {
+                    String eventName = "Timeout(tp=" + tp + ", key=" + key + ")";
+                    CoordinatorWriteEvent<Void> event = new CoordinatorWriteEvent<>(eventName, tp, coordinator -> {
+                        log.debug("Executing write event {} for timer {}.", eventName, key);
+
+                        // If the task is different, it means that the timer has been
+                        // cancelled while the event was waiting to be processed.
+                        if (!tasks.remove(key, this)) {
+                            throw new RejectedExecutionException("Timer " + key + " was overridden or cancelled");
+                        }
+
+                        // Execute the timeout operation.
+                        return new CoordinatorResult<>(operation.generateRecords(), null);
+                    });
+
+                    // If the write event fails, it is rescheduled with a small backoff except if the
+                    // error is fatal.
+                    event.future.exceptionally(ex -> {
+                        if (ex instanceof RejectedExecutionException) {
+                            log.debug("The delayed write event {} for the timer {} was not executed because it was " +
```

Comment:
I think the usage of delayed also confused me here. In theory if we never hit the timeout, we will repeatedly reschedule and hit this right?

Reply:
> I think the usage of delayed also confused me here. 

Let me remove `delayed` here. It is not necessary.

> In theory if we never hit the timeout, we will repeatedly reschedule and hit this right? And it steady state, we don't necessarily want to write the event.

I don't understand your point here. The timeout is only rescheduled if it fails with an unexpected exception. If the timeout is cancelled or if a new timeout with the same key is scheduled, the retry is aborted/ignored.

Reply:
Hmm. Maybe I am still misunderstanding the usage. Let's use an example:
> heartbeat operation which when expires we remove the member from the group

My understanding is that we schedule a heartbeat timeout and if we get the heartbeat, we reschedule the timeout. Is this not correct usage still?

Reply:
Yeah, that’s right. We reschedule the timeout on every heartbeat. So, unless the member fails, the timeout never expires.

Reply:
Ok -- so my original question is that in steady state, we expect to see this log message that we didn't execute the timer.

Reply:
Actually, no, we don’t because we expect the timer to be rescheduled before it expires and pushes the event to the queue. We would only see it when the timer expires and the event get pushed to the queue and the timer is rescheduled before the event is actually executed. This could for instance happen if the timeout event get right behind a heartbeat event in the queue.

Reply:
Oh because enqueue is in the run method 🤦‍♀️ my mental model of this is all turned around -- so we will just replace the  task in tasks if it gets rescheduled and only get to that exception + log if we made it to processing and a new task was scheduled. 👍